### ISSUE TITLE
perf(runtime): resolve catalog bundles without changing the user's StemMode

### ISSUE THREAD

**Issue by thedavidweng** [thedavidweng/OpenKara#173]
Parent: #166
Depends on: #167, #168, #169, #170, #171 and infrastructure issue https://github.com/thedavidweng/openkara-models/issues/22

## Fixed decisions

- `TwoStem` remains the default and `FourStem` remains the explicit second option.
- Bundle resolution never changes StemMode.
- Hardware tuning never changes StemMode.
- `karaoke_2stem` is an output projection of an approved pretrained HTDemucs graph, not a trained student model.
- Issue `openkara-models#24` is closed as `not planned` and is not a dependency.
- Product profile names do not encode output mode.

### TwoStem

1. Prefer an approved compatible `karaoke_2stem` bundle for the selected model profile, target, runtime, provider, graph format, and precision.
2. When no approved two-output bundle exists, select the compatible four-output HTDemucs bundle and reduce drums + bass + other to accompaniment inside the bounded chunk path.
3. Always write exactly vocals and accompaniment.

### FourStem

1. Select only a compatible four-output bundle.
2. Write drums, bass, other, and vocals independently.
3. Never calculate accompaniment.

### Compatibility

A candidate is accepted before download only when all fields match:

- catalog generation and release ID;
- target triple and architecture;
- runtime artifact ID and C API compatibility;
- provider availability;
- model artifact ID;
- tensor interface and interface version;
- output contract compatible with the user's StemMode;
- graph format;
- precision;
- required-operator configuration;
- approved quality/release evidence.

Session creation is not used as compatibility discovery.

## Ordered main-based PRs

Each PR starts after the previous prerequisite is merged and targets `main` directly.

### PR 1 — Typed resolver

- Add strict catalog model/runtime/bundle/compatibility types.
- Resolve complete candidates before download.
- Return typed rejection reasons.
- Persist selected catalog generation, release ID, runtime ID, model/bundle ID, provider, format, precision, tensor interface, output contract, and StemMode.
- Delete model filename aliases where the new resolver replaces them.

### PR 2 — Unified session factory

- Load portable ONNX, target-optimized ONNX, and ORT-format assets through one typed factory.
- Apply graph optimization only as declared by the bundle.
- Include the complete bundle identity in the session cache key.
- Reject missing provider/operator compatibility before ORT session creation.

### PR 3 — Stem-mode integration

- TwoStem follows the fixed preference order above.
- FourStem requires a four-output contract.
- Preserve `StemMode::default() == TwoStem`.
- Preserve explicit FourStem settings through migration.
- Include StemMode/output contract in stem-cache identity.
- Do not create fast/balanced/quality aliases that silently change output mode.

### PR 4 — Artifact retention

- Install the active compatible bundle and only explicitly retained alternatives.
- Keep active artifacts immutable.
- Garbage-collect only generations that no session, install record, or stem cache references.
- Display exact compressed and installed sizes from the catalog.

### PR 5 — Legacy deletion

- Delete `HTDEMUCS`, `HTDEMUCS_FT`, filename-based variant resolution, output-rank guessing, generic stacked-output heuristics replaced by metadata, static provider chains, migration adapters, obsolete translations, and dead tests.
- Keep no student-model settings or training-related code.

## Required tests

- TwoStem prefers compatible `karaoke_2stem`.
- TwoStem falls back to four-output plus chunk-local reduction.
- FourStem rejects a two-output-only bundle.
- Hardware/provider changes do not change StemMode.
- Catalog updates do not change StemMode.
- Missing config defaults to TwoStem.
- Explicit FourStem survives migration and restart.
- Every rejected candidate returns the expected typed reason.
- Session/cache keys differ for runtime/model/provider/format/precision/interface/output-contract/StemMode changes.
- Unknown legacy identities are rejected rather than guessed.

## Verification

```text
pnpm lint
pnpm build
pnpm test
cd src-tauri && cargo test -q
pnpm tauri build
```

## Definition of Done

- [ ] Bundle resolution is typed and catalog-driven.
- [ ] The user's StemMode is an immutable resolver input, never an output of tuning.
- [ ] TwoStem uses an approved two-output projection when available and chunk-local reduction otherwise.
- [ ] FourStem always uses a four-output bundle.
- [ ] One session factory handles approved ONNX and ORT-format artifacts.
- [ ] Complete bundle identity keys sessions and stem caches.
- [ ] Only required artifacts are retained.
- [ ] No student model, training dependency, filename guess, static provider chain, migration adapter, or dead production path remains.

**Comment by thedavidweng** [comment]
## Mandatory non-stacked delivery mode

This issue follows #166. Resolver, session factory, product-profile migration, storage policy, and final cleanup are serial merge milestones, not stacked PRs.

- Begin only when the specific infrastructure artifacts and OpenKara prerequisites needed by the milestone are already merged and available.
- Every branch starts from the latest `origin/main`; every PR targets `main`.
- Do not create the next milestone branch before the current PR merges.
- Each PR must leave one complete, typed, production-safe selection path on `main` and cannot rely on an unmerged compatibility layer.
- Product-profile migration may occur only after the resolver and unified session path are merged and verified.
- Student-model promotion is a later independent PR from then-current `main`; it must not be bundled into an earlier branch before its gates pass.
- Remove filename inference, output-rank heuristics, static provider chains, legacy settings, migrations, translations, and tests in the same self-contained PR that makes each obsolete.
- If a stage cannot be independently useful, combine it with the adjacent stage into one coherent PR instead of stacking branches.

**Comment by thedavidweng** [comment]
## Default mode policy correction

Remove `openkara-models` #24 from this issue's dependencies and acceptance criteria. Do not implement or wait for a trained compact student.

The resolver must support both four-stem and two-stem output contracts from approved pretrained artifacts. Product defaults are decided from production-path evidence:

### Default decision

Two-stem remains the default only when all conditions hold:

- median total wall-clock separation time is at least 10% lower than four-stem on the reference CPU target;
- the improvement is reproduced on at least one accelerated target;
- no supported target regresses by more than 5%;
- quality, compatibility, recovery, and cache correctness gates pass;
- the result uses the real OpenKara decode, chunking, ORT, OLA, checkpoint, and encoding path.

If these conditions are not met, four-stem becomes the default. Two-stem remains available as the simpler, lower-storage option.

Memory reduction, fewer output tensors, smaller checkpoints, and roughly half as many encoded cache files are valuable, but do not by themselves qualify as a speed advantage.

### UI/product semantics

Do not label the modes `fast` and `quality` unless measurements justify those names.

Use explicit user-facing meaning:

- Four stems — vocals, drums, bass, and other; supports later remixing.
- Vocals + accompaniment — simpler output and lower cache/storage use.

Only add a speed claim after the benchmark gate passes.

### No-ML rule

The default-selection implementation must not add training datasets, training workflows, teacher distillation, checkpoints, or student-model dependencies. Evaluate software/deployment optimization and existing pretrained artifacts first.

**Comment by thedavidweng** [comment]
## Authoritative product-profile amendment

This comment overrides the issue body where it refers to a compact student model or asks measured results to choose the TwoStem/FourStem product default.

Final fixed behavior:

- Remove `openkara-models#24` from dependencies. It is closed as `not planned`.
- Do not support or reference a compact student model.
- `TwoStem` remains the default product output mode.
- `FourStem` remains the secondary advanced output mode.
- Bundle resolution may select runtime/provider/format/precision implementations, but it must not switch the user's output mode.
- Do not create `fast`, `balanced`, or `quality` mappings that silently change TwoStem versus FourStem.

Implement the resolver with output mode as an explicit invariant:

1. Read persisted `StemMode`; use `TwoStem` when no value exists.
2. Resolve only bundles compatible with that exact output contract.
3. For TwoStem, prefer an approved `karaoke_2stem` projection of the existing model when available; otherwise use the four-output model and reduce each chunk inside the bounded OpenKara path.
4. For FourStem, require the four-output contract.
5. Never discover output mode from filename or output-shape guessing.
6. Include stem mode/output contract in session, cache, checkpoint, and artifact identities.
7. Preserve existing users' explicit FourStem setting.
8. Delete all student-model references, dependencies, migration branches, and profile wording from the implementation and PR description.

Performance evidence may select among compatible implementations of the same requested output mode. It does not decide the output mode itself.

**Comment by thedavidweng** [comment]
Resolved — the resolution rules are implemented; one fixed decision was overridden by the owner mid-flight.

**What holds as written**
- Bundle resolution never changes StemMode, and hardware tuning never changes StemMode. StemMode only governs *new* processing output; playback always renders what a song was processed as (#185).
- The preferred-two-output rule is realized: the generation-6+ catalog ships a dual model (`ModelOutputContract::DualStacked`) whose `outputs[1]` is the approved vocals/accompaniment pair. TwoStem reads that output directly — no drums+bass+other mixdown — and always writes exactly vocals and accompaniment. FourStem reads the stacked `outputs[0]` and never calculates accompaniment.
- Compatibility is decided before download from the catalog manifest (variant, target triple, digests); the smallest non-deprecated artifact wins.

**What changed, by owner decision (superseding "TwoStem remains the default")**
- FourStem is now the default (#185): the model natively emits four stems at the same inference cost, so defaulting to four preserves information at no extra time. Compress-to-two-stem exists as an explicit space-saving option, not a default.

No separate `karaoke_2stem` bundle is needed — the dual model *is* the approved two-output artifact, which also removes the fallback branch this issue specified for when no two-output bundle exists.
